File: 3. 无重复字符的最长子串.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const map = {}; // symbol: index;
    let left = 0;
    let maxLength = 0;
    for (let right = 0; right < s.length; right++) {
        const nowSymbol = s[right];
        if (map.hasOwnProperty(nowSymbol) && map[nowSymbol] >= left) {
            // map中的记录不需要清除，只要判断重复字符是否在left-right的范围中
            left = map[nowSymbol] + 1;
        }
        map[nowSymbol] = right;
        maxLength = Math.max(right - left + 1, maxLength);
    }
    return maxLength;
};

File: 3. 无重复字符的最长子串.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> m;
        int left = 0;
        int maxLength = 0;
        for (int right = 0; right < s.length(); right++) {
            char nowSymbol = s[right];
            if (m.find(nowSymbol) != m.end() && m[nowSymbol] >= left) {
                left = m[nowSymbol] + 1;
            }
            m[nowSymbol] = right;
            maxLength = max(right - left + 1, maxLength);
        }
        return maxLength;
    }
};